### ISSUE TITLE
[feature] clash-avoiding naming of instance variables on elements

### ISSUE THREAD

**Issue by trusktr** [ryansolid/solid#319]
The output currently does things like

```js
    _el$.__foo = () => {};
```

but there's a high chance of that colliding with properties on custom elements. Example (pate into the template explorer):

<details><summary>Input:</summary>

```tsx
class SomeEl extends HTMLElement {
  private __foo = {/*...some payload...*/}
  
  connectedCallback() {
    setTimeout(() => {
      this.dispatchEvent(new CustomEvent('foo', {
        detail: this.__foo
      }))
    }, 1000)
  }
}

customElements.define('some-el', SomeEl)

class OtherEl extends HTMLElement {
  template = () => (
    <>
      Test:
      <some-el ref={this.el} onfoo={payload => console.log(payload)} />
    </>
  )
  
  connectedCallback() {
    this.append(this.template())
  }
}

customElements.define('other-el', OtherEl)

document.body.append(document.createElement('other-el'))
```

</details>

I couldn't get that demo to work in template explorer, but I didn't spend too much time. The template clone result appears to have an `undefined` value.

But as you can see in the Output tab, the template output will overwrite the `some-el` element's private `__foo` property, and this won't be clear to the end user why it happens.

<details><summary>Output:</summary>

```js
import { template, delegateEvents, currentContext } from 'solid-js/web';

const _tmpl$ = template(`<some-el></some-el>`, 2);

// source: counter.tsx
class SomeEl extends HTMLElement {
  __foo = {
    /*...some payload...*/
  };

  connectedCallback() {
    setTimeout(() => {
      this.dispatchEvent(new CustomEvent('foo', {
        detail: this.__foo
      }));
    }, 1000);
  }

}

customElements.define('some-el', SomeEl);

const _self$ = undefined;

class OtherEl extends HTMLElement {
  template = () => ["Test:", (() => {
    const _el$ = _tmpl$.cloneNode(true);

    _el$.__foo = payload => console.log(payload); // <--------------------------------- OOPS

    const _ref$ = _self$.el;
    typeof _ref$ === "function" ? _ref$(_el$) : _self$.el = _el$;
    _el$._context = currentContext();
    return _el$;
  })()];

  connectedCallback() {
    this.append(this.template());
  }
}

customElements.define('other-el', OtherEl);
document.body.append(document.createElement('other-el'));

delegateEvents(["foo"]);

```

</details>

**Comment by trusktr** [comment]
Idea, change it to this:

```js
const foo = Symbol('foo')
// ...
delegateEvents([foo]);
```

**Comment by ryansolid** [comment]
Right event delegation syntax.. I could add another special character in there. I basically went with __ as sort of artificial placeholders for `on`. But really it could be anything. I just need to write a property per event name to the element so that when we walk the tree we can find the appropriate information.

Is it just that `__` is common for private variables? I'd prefer to not use symbols for performance issues and that I want them to be the same across files/compilations etc.. like you might delegate click in each file and I need it to be the same event name.. I take the part after `__` and do an `addEventListener` with it, if it isn't already in the Set. So it might be called with `click` a number of places including in spread operators, places that aren't precompiled.

**Comment by trusktr** [comment]
> Is it just that __ is common for private variables

Yeah, I use that convention, and I've seen it in other places (people coming from python for example are accustomed to `__` defining an actually-private variable and it isn't just naming convention).

**Comment by ryansolid** [comment]
I can look at maybe `$$` I suppose there is always a chance of a collision.

**Comment by trusktr** [comment]
That's probably a lot better (underscores for privacy are a lot more common).

Couple other idea just in case:

- store symbols on an object that the template imports from solid-js, so there's only ever one `foo` symbol.
- use a Map or WeakMap, map elements to event functions.

```js
import {eventMap} from 'solid-js'

eventMap.get('foo').set(_el$, theFunction) // probably not as fast as strings
```

**Comment by ryansolid** [comment]
Fixed in 0.24.0.
